File: ios/Arculus/CSDK/WalletSelectWallet.swift
import CSDK

enum ApplicationAID {
    case v1
    case v2
    
    public var aid: [UInt8] {
        switch self {
        case .v1: return [0x4a, 0x4e, 0x45, 0x54, 0x5f, 0x4c, 0x5f, 0x01, 0x01, 0x57]
        case .v2: return [0x41, 0x52, 0x43, 0x55, 0x4C, 0x55, 0x53, 0x01, 0x01, 0x57]
        }
    }
}

enum WalletSelectWalletError: Error, LocalizedError {
    case invalidAID
    case expectedAIDNotMet(ApplicationAID)
    
    var errorDescription: String? {
        switch self{
        case .invalidAID:
            return "Invalid application AID"
        case let .expectedAIDNotMet(applicationAID):
            return "Expected \(applicationAID) application AID not met"
        }
    }
}

final class WalletSelectWallet: CSDKAPICall<[UInt8]> {
    private var applicationAID: ApplicationAID
    
    init(wallet: OpaquePointer, applicationAID: ApplicationAID) {
        self.applicationAID = applicationAID
        
        super.init(wallet: wallet)
    }
    
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletSelectWalletRequest(wallet, applicationAID.aid, &len) else {
            throw CSDKAPICallError(.requestCreationFailed("wallet select"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws -> ResponseType {
        guard let result = WalletSelectWalletResponse(wallet, bytes, bytes.count) else {
            throw CSDKAPICallError(.responseParsingFailed("wallet select"))
        }
        
        let data = Data(bytes: result.pointee.ApplicationAID, count: size_t(result.pointee.ApplicationAIDLength))
        
        let aid = [UInt8](data)
        
        return aid
    }
}

File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletSelectWallet.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class InvalidAIDException : Exception("Invalid application AID")

class ExpectedAIDNotMetException(expectedAID: WalletSelectWallet.ApplicationAID) :
  Exception("Expected ${expectedAID.aid} application AID not met")

class WalletSelectWallet(
  wallet: Pointer,
  private val applicationAID: ApplicationAID
) : CSDKAPICall<ByteArray>(wallet) {
  enum class ApplicationAID(val aid: ByteArray) {
    V1(byteArrayOf(0x4a, 0x4e, 0x45, 0x54, 0x5f, 0x4c, 0x5f, 0x01, 0x01, 0x57)),
    V2(byteArrayOf(0x41, 0x52, 0x43, 0x55, 0x4C, 0x55, 0x53, 0x01, 0x01, 0x57))
  }

  override suspend fun request(): Array<ByteArray> {
    val len = SizeTByReference()

    val pointer = CSDK.WalletSelectWalletRequest(wallet, applicationAID.aid, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray): ByteArray {
    val len = SizeTByReference()

    val result = CSDK.WalletSelectWalletResponse(
      wallet,
      bytes,
      bytes.size
    )

    val sAID = CSDK.WalletGetApplicationAID(result, len)

    return sAID.getByteArray(0, len.toIntChecked())
  }
}
